### PR TITLE
Store: Show a customs form if the order requires it to be shipped

### PR DESCRIPTION
**Pull request by DanReyLop**
This PR adds a new step to the "Purchase Label" modal, called `Customs`. It's displayed between "Packages" and "Rates". It's only displayed if Customs information is needed to get the package from `Origin` to `Destination`. The rules are as follows:
- Customs is required when sending a package to a different country or territory, except if the shipment is between the "proper" United States, Puerto Rico, and the Virgin Islands.
- Shipping from or to a military base (represented by the special `AA`, `AE` and `AP` states) requires Customs.

The `add/intl-labels` branch in `woocommerce-services` removes the "Destination country" restriction for showing the "Create Label" button on WP-Admin, and also ensures that the REST API replies with **all** the countries in the world, not only the "supported" ones. That means that, when this is released, your `woocommerce-services` plugin is outdated, and you receive an order from a previously-unsupported country, you'll see this:
screen shot 2018-04-02 at 12 27 05

Not pretty, but it's the simplest way to ensure backwards-compatibility.

If you have the `add/intl-labels` branch on `woocommerce-services`, you'll be able to set the `Destination` country to anywhere in the world, and these "countries" as `Origin`:
screen shot 2018-04-02 at 12 28 18
Those are the "countries" (as listed in WooCommerce) that have at least 1 open USPS office. All of them use the same `ZIP` code format as continental US, so there's no problem supporting them.

~I added a dummy checkbox in the Customs step so you can check the flow even though the contents aren't there yet. The step is considered valid if the checkbox is selected, and erroneous otherwise. Some things to consider:~
~- The Customs form only depends on the `Packages` information. In fact, ideally it would be better displayed as part of the `Packages` step, if there's a way to do it without overcrowding it (I doubt it).~
~- The Customs information will never affect the calculated Rates.~

### CODE REVIEW

**Comment by kellychoffman** [comment]
I tested this PR using the calypso.live PR and got a few errors. Let me know if any of the below is expected:

I tested this out on a new order (created manually) and when I went to fulfill the order, I got this error (global error on page): `Error validating destination address: Server could not read response.`. (I didn't have a destination address entered) I also had this error on the Destination address card: `You've edited the address, please revalidate it for accurate rates`. No matter what I enter for destination address, I get the `Error validating destination address: Server could not read response.`global page error.

- Is it necessary to have a page error when first opening the modal? Can we just have an error state on the destination address card? 
- Can we change the copy of the Destination address card to not mention editing the address since I didn't edit anything yet?

**Comment by DanReyLop** [comment]
> I tested this out on a new order (created manually) and when I went to fulfill the order, I got this error (global error on page): Error validating destination address: Server could not read response.. 

That's just a generic HTTP error returned from the WordPress site. In this case, it probably means that the `api.woocommerce.com` server isn't running, so it just fatals when trying to validate the address. That's what's causing the `You've edited the address` error too.

Please check if you have the constant `WOOCOMMERCE_CONNECT_SERVER_URL` set to something in your site.

File: client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/customs-step/package-row.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+					id={ packageId + '_contentsType' }
+					title={ translate( 'Contents type' ) }
+					value={ contentsType || 'merchandise' }
+					updateValue={ props.setContentsType }
+					valuesMap={ {
+						merchandise: translate( 'Merchandise' ),
+						documents: translate( 'Documents' ),
+						gift: translate( 'Gift' ),
+						sample: translate( 'Sample' ),
+						other: translate( 'Other…' ),
```
Comment (a8ci18n) [inline client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/customs-step/package-row.js]: Hi! I've found a possible matching string that has already been translated **71** times:
`translate( 'Other' )` *ES Score: 12*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by DanReyLop** [comment]
This PR now includes the entire feature. Here's the information dump.

This PR requires https://github.com/Automattic/woocommerce-connect-server/pull/1074 in the server to actually purchase labels, but the entire flow until the "Rates" step will work without it.

In https://github.com/Automattic/wp-calypso/pull/23829/commits/23b92536ec0106d415da0f85a76b93fe3c14d257, I added a feature flag so it's disabled in `production`. With the feature disabled:
- You won't be able to open the "Purchase Label" flow if a Customs form is needed to send from the store address to the order shipping address. That means that only orders between the US, Puerto Rico and the Virgin Islands will be able to print labels.
- Once the Purchase Label modal is open, you'll still be able to change the origin or destination country, but only to US, Puerto Rico or Virgin Islands.

With the feature flag enabled, but a version of the plugin that doesn't include https://github.com/Automattic/woocommerce-services/pull/1382:
- If either the origin or destination country is not US or Puerto Rico, the Label flow will work, but in the origin/destination country dropdown you'll only see as options US, Puerto Rico, and the country code you're shipping from/to.
- The HS Tariff Code, item descriptions, and item origin countries won't be remembered between orders, because the PHP code to store them isn't present.
- Everything else, including actually purchasing a label, should work. Not everyone updates their plugins immediately so I coded this with backwards-compatibility in mind.

The Customs form is needed in order to get rates, because of the way our EasyPost integration works (the Shipment object needs to be created before getting rates), so the Rates step won't be updated unless the Customs form is valid.

The Customs step UI is in `client/extensions/woocommerce/woocommerce-services/views/shipping-label/label-purchase-modal/customs-step/*`.

I've linked a demo of the feature in the master thread.

**Comment by kwight** [comment]
Hm, this PR is now +1000 LOC; is the idea to break it into multiple PRs? The Redux stuff (actions, reducers, etc) can be broken out, then UI tested and merged after. Also each PR could use solid testing instructions for those of us unfamiliar with all the moving parts 🙃

File: client/extensions/woocommerce/woocommerce-services/state/shipping-label/selectors.js [hunk 2]
```diff
@@ -260,9 +299,24 @@ export const getCustomsErrors = ( packages, customs ) => {
 				errors.restrictionExplanation = translate( 'Please describe what kind of restrictions this package must have' );
 			}
 
-			const totalValue = sumBy( pckg.items, ( { quantity, product_id } ) => quantity * customs.items[ product_id ].value );
-			if ( totalValue > 2500 && ! pckg.itn ) {
-				errors.itn = translate( 'ITN is required for shipments over $2,500' );
+			const classesAbove2500usd = new Set();
+			forEach( pckg.items, ( { product_id } ) => {
+				const { tariffNumber } = customs.items[ product_id ];
+				if ( 45 < valuesByTariffNumber[ tariffNumber ] ) {
```
[the remark this answers is not in the document]
Comment (DanReyLop) [reply client/extensions/woocommerce/woocommerce-services/state/shipping-label/selectors.js]: Oops, I messed this up during testing. It's supposed to be `2500`. Fixed.

**Comment by DanReyLop** [comment]
@kwight Sorry for not responding earlier. Feedback for this PR is mainly being collected in p7bbVw-2o2-p2, so it looks abandoned here but it really isn't. I've done my best to split it (into 3) but it's probably still not small enough for Calypso standards. I'll do my best from now own to submit smaller PRs but for this one I think it would be more trouble than it's worth.

> what do you think about splittting this PR into smaller chunks?

@v18 Done.
Part 1: https://github.com/Automattic/wp-calypso/pull/24976
Part 2: https://github.com/Automattic/wp-calypso/pull/24975
Part 3 (still big, but not as scary): https://github.com/Automattic/wp-calypso/pull/24974

I'm closing this one since it's no longer relevant.

**Pull request closed without merging**